File: Data Structures & Algorithms/validate-parentheses/submission-2.java
class Solution {
    public boolean isValid(String s) {
        Map<Character,Character> map = new HashMap<>();
        map.put(')','(');
        map.put('}','{');
        map.put(']','[');
        
        Stack<Character> charStack = new Stack<>();

        for(Character c: s.toCharArray()){
            if(map.containsKey(c)){
                if(!charStack.isEmpty() && map.get(c).equals(charStack.peek())){
                    charStack.pop();
                }
                else{
                    return false;
                }        
            }
            else{
                    charStack.push(c);
                }
        }
        return charStack.isEmpty();
    }
}

File: Data Structures & Algorithms/asteroid-collision/submission-2.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for a in asteroids:
            while stack and a < 0 and stack[-1] > 0:
                if stack[-1] + a == 0:
                    stack.pop()
                    a = 0
                    break
                elif stack[-1] + a < 0:
                    stack.pop()
                else:
                    a = 0
            if a:
                stack.append(a)
        return stack